### ISSUE TITLE
Escape perm whitelist using symlink

### ISSUE THREAD

**Issue by kevinkassimo** [denoland/deno#2318]
e.g. Currently creating a symlink under write-allowed directory to other directories would allow escaping the perm rules and access any file.

**Comment by piscisaureus** [comment]
I agree this is a problem, unfortunately it's very hard to fix (on some platforms at least).
On OS X and BSD derivatives it'd be possible to query the filename of an open file using fcntl(F_GETPATH). 
On windows, something similar is possible using NtQueryInformationFile, although this is a bit tricky to implement since the winapi crate doesn't expose these functions.
On linux, I'm sure it's possible somehow, but I don't know how...

This is of course for the case of *opening* files. For renaming, deleting, moving etc the trick would be to rely on calling `openat()` and after that checking that the path is still inside a permitted subdirectory, and then performing the move/delete operation on the file descriptor using `renameat()/unlinkat()`. 

So this is definitely fixable, but quite a but of work. That said, I think it'd be worthwhile and important to do it.

**Comment by nayeemrmn** [comment]
Bump. This should really be a 1.0 blocker unless it's clarified that 1.0 =/= production-ready.

@bartlomieju Permissions label.

**Comment by dubiousjim** [comment]
@piscisaureus Yeah I haven't yet found out how to recover a path from an open fd on Linux either. I don't know if it's possible (short of searching the device for the inode... or parsing a file in `/proc`, which wouldn't be _as_ ridiculously expensive, but still too expensive).

A general problem here is that an open fd _needn't have_ any existing path to it. Isn't it a common strategy for handling temp files to open the file and then unlinking its path (or that of the directory that contains it), while keeping the fd still open.

**Comment by dubiousjim** [comment]
Related issues: #4437, related PRs: #2317, #2319.
Cc: @kevinkassimo @ry @piscisaureus @bartlomieju @nayeemrmn @afinch7 

Does the following sound reasonable for a minimal fix? (We might look for a more ambitious fix later.)

1. Since `op_link` will have to deposit its destination file in a directory with write permissions, thus exposing the original to write access, we require write access on this operation's source file (Currently we only require read access... it's not clear that's even necessary, unless the destination directory also has read access. If one's insisting on read access because this operation at least verifies the existence of the original file, we're already pretty unsystematic about this --- follow up at #4423.)

2. Since `op_symlink` will have to deposit its destination file in a directory with write permissions, thus exposing the original to write access, we require that either (a) this operation's source file also has write access, or (b) this operation's source file doesn't exist. Also, if the destination file will have read access, we require that the source file also have read access, if it exists. (Currently we don't do any access or existence checks on the source file.)

These restrictions are still sub-optimal because they don't protect against funny business if a script is run with different permissions at different times, or if some part of the filesystem is only mounted after the script has run once. But they should block straightforward exploits we're vulnerable to now.

**Comment by kevinkassimo** [comment]
@dubiousjim I think the proposal is reasonable as a stopgap, while it is not perfect, it is way better than what we have atm. Essentially
1. `op_link` requires a superset of permissions on link target compared to that of link file.
2. `op_symlink`, similar to `op_link`, but I do feel debatable whether part (b) is safe: it is an attack for the script to create a symlink beforehand, waiting for a sensitive log to appear at that location pointed to, and try reading contents. We might as well ban creating symlinks pointing to a non-existent file, unless the current permissions implies that that once this link target is created, will have superset of permissions than that of link file itself.

**Comment by Spoonbender** [comment]
I think this is much better, but still not bullet-proof, and I fear the rabbit hole is deep.

One thing that comes to mind is relative symlinks that lead to a parent directory.
The ability to move such a link, either directly or indirectly by moving its containing directory, leads to additional holes.

Consider the following exploit:
```ts
import { move, ensureSymlink, ensureDir, walkSync  } from "https://deno.land/std/fs/mod.ts";
await ensureDir("/sandbox/temp");
await Deno.chdir("/sandbox/temp");
await ensureSymlink("..", "/sandbox/temp/exploit");
await move("/sandbox/temp/exploit", "/sandbox/exploit");
for (const entry of walkSync("/sandbox/exploit/etc")) {
    console.log(entry.path);
}
```
Run the script and provide it read and write access to /sandbox. The script breaks out of the sandbox.

**Comment by Spoonbender** [comment]
The more thought I give this issue, the more complex it seems.

The O/S lets the users create symlinks to wherever they want to, and there is no security risk in that - since the O/S always verifies actions against the ACL of the true target. 
OTOH, Deno doesn't verify that ops are allowed on the true target. Instead, it verifies that ops are allowed on the "symbolic" target, and also attempts to ensure that the script doesn't modify the FS in a way that creates new symbolic paths that break out of the allowlist. This approach, however, seems quite fragile and prone to various attacks (which I can elaborate on).

If we don't intend for Deno to utilize O/S mechanisms for enforcing permissions on the target (ACLs, chroot etc.), then I think our best bet is to verify ops against the true target. This means that when an op is initiated against some path, we need to resolve all symlinks along this path and get a hold of a file descriptor, verify that it is in the allowlist, then operate on this fd. Unfortunately, this resolution process needs to be conducted for every call - otherwise we're prone to attacks that exploit the cached result. (The resolution process can be skipped as long as fine-grained filesystem permissions are not used).

**Comment by benatkin** [comment]
If only existing symlinks allow it to escape the directory, I think it works for most situations.

I tried creating a symlink dynamically inside of Deno from the directory I could write to, to a file in another directory, and Deno wouldn't do that. So I think it might not be an issue most of the time.

**Comment by nayeemrmn** [comment]
Fixed by #12554.
